### ISSUE TITLE
UnobservedTaskException after Client.Disconnect()

### ISSUE THREAD

**Issue by nixtar** [TalAloni/SMBLibrary#246]
Hello,

I think this might be related/similar to #243.

I have noticed that after calling `client.Disconnect()` there is a Socket related exception raised somewhere and ends up bubbling up under TaskScheduler.UnobservedTaskException event.

```
System.Net.Sockets.SocketException (995): The I/O operation has been aborted because of either a thread exit or an application request.
   at System.Net.Sockets.Socket.AwaitableSocketAsyncEventArgs.ThrowException(SocketError error, CancellationToken cancellationToken)
   at System.Net.Sockets.Socket.AwaitableSocketAsyncEventArgs.System.Threading.Tasks.Sources.IValueTaskSource<System.Int32>.GetResult(Int16 token)
   at System.Threading.Tasks.ValueTask`1.ValueTaskSourceAsTask.<>c.<.cctor>b__4_0(Object state)
```

We have an app that needs to utilise the UnobservedTaskException event to capture async GUI related exceptions with the UI Framework its using.

I tried to reproduce this using the socket example [here ](https://learn.microsoft.com/en-us/dotnet/api/system.net.sockets.socket?view=net-8.0#examples) and calling `Disconnect(false)` and `Close()` the same way this library does and the exception is not raised. 

While looking at this I found the following remark about BeginReceive:
https://learn.microsoft.com/en-us/dotnet/api/system.net.sockets.socket.beginreceive?view=net-8.0
```
Close the Socket to cancel a pending BeginReceive.  
When the Close method is called while an asynchronous operation is in progress, the callback provided to the BeginReceive method is called.  
A subsequent call to the EndReceive method will throw an ObjectDisposedException (before .NET 7) or a SocketException (on .NET 7+) to indicate that the operation has been cancelled.
```
Of note here: **SocketException (on .NET 7+) to indicate that the operation has been cancelled.**

I also note in the doco that `The asynchronous BeginReceive operation must be completed by calling the EndReceive method`.
With that in mind perhaps EndReceive needs to be called before Close while disconnecting?

The following can reproduce this in a console app:

```c#
class Program
{
    static void Main(string[] args)
    {
        TaskScheduler.UnobservedTaskException += (sender, eventArgs) =>
        {
            var ex = eventArgs.Exception;
            throw ex;
        };
        
        var creds = new NetworkCredential(blah blah);
        var text = ReadAllText(@"\\server\textfile.txt", creds);
        Console.WriteLine(text);
        Console.WriteLine("Wait 1 second!");
        Thread.Sleep(1000);
        GC.Collect(); // Force GC to reproduce the issue consistently, UnobservedTaskException only fire when the GC "collects" them.
        Console.WriteLine("Done!");
        Console.ReadKey();
    }
    
    static string ReadAllText(string path, NetworkCredential? credential = null)
    {
        // Parse the SMB path to extract server name and share path
        Uri smbUri = new Uri(path);
        string serverName = smbUri.Host;
        string shareName = smbUri.Segments[1].TrimEnd('/');
        string filePath = string.Join("", smbUri.Segments, 2, smbUri.Segments.Length - 2).Replace("/", "\\");

        // Initialize SMB client
        SMB2Client client = new SMB2Client();
        bool isConnected = client.Connect(serverName, SMBTransportType.DirectTCPTransport);

        if (!isConnected)
        {
            throw new Exception("Could not connect to the SMB server.");
        }

        try
        {
            NTStatus status;
            if (credential != null)
            {
                // Authenticate with provided credentials
                status = client.Login(credential.Domain, credential.UserName, credential.Password);
                if (status != NTStatus.STATUS_SUCCESS)
                {
                    throw new Exception("SMB login failed.");
                }
            }

            // Connect to the share
            object shareObject;
            ISMBFileStore fileStore = client.TreeConnect(shareName, out status);
            if (status != NTStatus.STATUS_SUCCESS)
            {
                throw new Exception("Could not access the SMB share.");
            }

            // Open the file
            object fileHandle;
            FileStatus fileStatus;
            status = fileStore.CreateFile(out fileHandle, out fileStatus, filePath, AccessMask.GENERIC_READ, 0, ShareAccess.Read, CreateDisposition.FILE_OPEN, CreateOptions.FILE_NON_DIRECTORY_FILE, null);
            if (status != NTStatus.STATUS_SUCCESS)
            {
                throw new Exception("Could not open the file.");
            }

            StringBuilder fileContents = new StringBuilder();
            int bufferSize = 4096; // Adjust buffer size as needed
            byte[] buffer;
            long offset = 0; // Start at the beginning of the file
            bool endOfFileReached = false;
        
            while (!endOfFileReached)
            {
                status = fileStore.ReadFile(out buffer, fileHandle, offset, bufferSize);
                if (status == NTStatus.STATUS_SUCCESS)
                {
                    if (buffer != null && buffer.Length > 0)
                    {
                        fileContents.Append(Encoding.UTF8.GetString(buffer, 0, buffer.Length));
                        offset += buffer.Length; // Move the offset forward
                    }
                }
                else if (status == NTStatus.STATUS_END_OF_FILE)
                {
                    endOfFileReached = true;
                    _ = fileStore.CloseFile(fileHandle);
                }
                else
                {
                    _ = fileStore.CloseFile(fileHandle);
                    throw new Exception("Could not read the file, unexpected NTStatus: " + status);
                }
            }
            
            // Return the file contents as a string
            return fileContents.ToString();
        }
        finally
        {
            // Cleanup: logoff and disconnect
            if (client is { IsConnected: true })
            {
                client.Logoff();
                client.Disconnect();
            }
        }
    }
}
```

**Comment by TalAloni** [comment]
Thanks Nick,
Did this issue occur on 1.5.1.2?

**Comment by nixtar** [comment]
Sorry for the delay,

Yes, this is on 1.5.1.2. I just tested now on 1.5.1.1 and the exception is not thrown.
This makes sense however as Close is not being called.

For now as a workaround, I've added a check for this particular exception in our `UnobservedTaskException` handler and am ignoring it.

**Comment by TalAloni** [comment]
Thanks Nick,
I've spent some time on this,
Using the example you provided, I can see the exception being thrown in a thread pool thread in .NET 6.0.
Interestingly, this does not happen when running under .NET Framework 2.0 and .NET Framework 4.0.
After trying various code changes, it's not clear to me what I can do (within reason) to avoid that behavior.
It's noteworthy that even without calling Socket.Close I'm observing that behavior. (Disconnect triggers it as well)

Given that this is a logging issue rather than an operational bug, and given that I believe I'm following the documentation, I'm closing this issue.
If you have a suggestion on how to address this issue, I'll gladly look into it.

**Comment by nixtar** [comment]
Yeah that's fair enough, I also can't see a way forward without possibly refactoring all the socket handling and likely breaking compatibility with older dotnets. (eg not using Begin* methods)

Thanks for your efforts regardless. 👍

**Comment by TalAloni** [comment]
I believe this issue has been resolved in v1.5.2
